File: Code/BinarySearch/Search_a_2D_Matrix_II_240.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    bool searchMatrix(vector<vector<int>> matrix, int target)
    {
        /* start from first row last column
        then
        if its greater than taregt move downward ++i
        else move left --j;
        */
        int i = 0;
        int j = matrix[0].size() - 1;
        while (i < matrix.size() && j >= 0)
        {
            if (matrix[i][j] == target)
                return true;
            else if (matrix[i][j] > target)
                --j;
            else
                ++i;
        }
        return false;
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto t = obj.searchMatrix({{1, 4, 7, 11, 15}, {2, 5, 8, 12, 19}, {3, 6, 9, 16, 22}, {10, 13, 14, 17, 24}, {18, 21, 23, 26, 30}}, 5);
    return 0;
}

File: java/BinarySearch/Search_a_2D_Matrix_II_240.java
package java.BinarySearch;

import java.util.*;
import java.io.*;

public class Search_a_2D_Matrix_II_240 {
    /* start from first row last column
        then 
        if its greater than taregt move downward ++i
        else move left --j;
        */
    public boolean searchMatrix(int[][] matrix, int target) {
        int i = 0;
        int j = matrix[0].length - 1;
        while (i < matrix.length && j >= 0) {
            if (matrix[i][j] == target)
                return true;
            else if (matrix[i][j] > target)
                --j;
            else
                ++i;
        }
        return false;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {

            // int x = ip.nextInt();
            // int y = ip.nextInt();
            // op.println(x + y);

            int grid[][] = { { 1, 4, 7, 11, 15 }, { 2, 5, 8, 12, 19 }, { 3, 6, 9, 16, 22 }, { 10, 13, 14, 17, 24 },
                    { 18, 21, 23, 26, 30 } };
            Search_a_2D_Matrix_II_240 ob = new Search_a_2D_Matrix_II_240();
            boolean ans = ob.searchMatrix(grid, 5);
            op.println(ans);
        }
    }
}
